### PR TITLE
mark --network-plugin-dir deprecated for kubelet

### LINKED ISSUE #43967: kubelet 1.6 looks for CNI binaries in wrong path
**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:36:33Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.0", GitCommit:"fff5156092b56e6bd60fff75aad4dc9de6b6ef37", GitTreeState:"clean", BuildDate:"2017-03-28T16:24:30Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: bare-metal
- **OS** (e.g. from /etc/os-release): Ubuntu 14.04
- **Kernel** (e.g. `uname -a`):  Linux dev-k8s-controller-p1 4.4.0-53-generic #74~14.04.1-Ubuntu SMP Fri Dec 2 03:43:31 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
- **Install tools**: chef cookbook
- **Others**: Docker 1.12.6, calico 2.1

CNI binaries:
```
/opt/cni/bin/
├── calico
├── calico-ipam
├── flannel
├── host-local
└── loopback
```

kubelet config:
```
 /usr/bin/kubelet \
    --register-schedulable=false \
    --pod-manifest-path=/etc/kubernetes/manifests \
    --node-labels=k8s-role=k8s-controller \
    --allow-privileged=true \
    --api-servers=https://dev.kubernetes-controller.service.consul:6443 \
    --cloud-provider= \
    --cluster-dns=10.223.0.10 \
    --cluster-domain=cluster.local \
    --container-runtime=docker \
    --docker=unix:///var/run/docker.sock \
    --network-plugin=cni \
    --network-plugin-dir=/etc/cni/net.d \
    --kubeconfig=/var/lib/kubelet/kubeconfig \
    --serialize-image-pulls=false \
    --tls-cert-file=/var/lib/kubernetes/kubelet-cert.pem \
    --tls-private-key-file=/var/lib/kubernetes/kubelet-key.pem \
    --v=2
```

**What happened**:
When trying to run pod error message is generated:
```
Events:
  FirstSeen	LastSeen	Count	From				SubObjectPath	Type		Reason		Message
  ---------	--------	-----	----				-------------	--------	------		-------
  9m		9m		1	kubelet, dev-k8s-controller-p1			Warning		FailedSync	Error syncing pod, skipping: failed to "CreatePodSandbox" for "fluentd-node-agent-kbrw4_kube-system(9813fecc-17df-11e7-9c75-5254009ef6db)" with CreatePodSandboxError: "CreatePodSandbox for pod \"fluentd-node-agent-kbrw4_kube-system(9813fecc-17df-11e7-9c75-5254009ef6db)\" failed: rpc error: code = 2 desc = NetworkPlugin cni failed to set up pod \"fluentd-node-agent-kbrw4_kube-system\" network: failed to find plugin \"loopback\" in path [/opt/loopback/bin /etc/cni/net.d]"

  9m	2s	45	kubelet, dev-k8s-controller-p1		Normal	SandboxChanged	Pod sandbox changed, it will be killed and re-created.
  9m	2s	45	kubelet, dev-k8s-controller-p1		Warning	FailedSync	Error syncing pod, skipping: failed to "KillPodSandbox" for "9813fecc-17df-11e7-9c75-5254009ef6db" with KillPodSandboxError: "rpc error: code = 2 desc = NetworkPlugin cni failed to teardown pod \"fluentd-node-agent-kbrw4_kube-system\" network: failed to find plugin \"calico\" in path [/etc/cni/net.d /opt/calico/bin]"
```

So it seems that kubelet is trying to find CNI binaries in paths:
- /opt/loopback/bin
- /etc/cni/net.d
- /opt/calico/bin

But it should look at `/opt/cni/bin` since this is the default valur for `--cni-bin-dir` parameter.

**What you expected to happen**:
Pod starts.

**How to reproduce it** (as minimally and precisely as possible):
Run pod:
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx
spec:
  containers:
  - image: nginx
    name: nginx
```

with kubelet 1.6 and CNI binaries in those path (default for calico):
```
/opt/cni/bin/
├── calico
├── calico-ipam
├── flannel
├── host-local
└── loopback
```

**Anything else we need to know**:
After copy `/opt/cni` to `/opt/calico` and `/opt/loopback` pod starts fine.

### PR DESCRIPTION
**Pull request by supereagle**
**What this PR does / why we need it**:

**Which issue this PR fixes** : fixes #43967

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by resouer** [comment]
@k8s-bot ok to test

**Comment by mtaufen** [comment]
Can you file an issue to track the deprecation of this flag and note the issue number in a comment next to the `MarkDeprecated` call? See https://github.com/kubernetes/kubernetes/issues/40229 as an example issue.

**Comment by freehan** [comment]
Thanks for the PR. 

There are some bootstrap scripts that are still using this flag. Can you also switch them to use `cni-bin-dir` and `cni-conf-dir`?

**Comment by supereagle** [comment]
@mtaufen  Good practice. Have created the issue #46410 to track this, and added the comment.

**Comment by supereagle** [comment]
@freehan  Of course, I will update these bootstrap scripts.

**Comment by supereagle** [comment]
According to the [comment](https://github.com/kubernetes/kubernetes/pull/43995#issuecomment-291261567):
> I am not sure how cni and kubenet share these flags. Below is my interpretation.
> * If `--network-plugin=cni`, `--cni-conf-dir` and `--cni-bin-dir` are used with their individual defaults.
> * If `--network-plugin=kubenet` is used, `--network-plugin-dir` is used.

The `--network-plugin-dir` is used for kubenet network, can this flag really be deprecated or removed?

**Comment by freehan** [comment]
@supereagle 
kubenet is also compatible with `cni-bin-dir`. 

If I remember correctly, `cni-bin-dir` preempts `network-plugin-dir`.

**Comment by supereagle** [comment]
@freehan You are right, thanks for your clarification. I have updated related bootstrap scripts.  PTAL.

As some test case code has been updated, kindly ping @Random-Liu .

File: hack/make-rules/test-e2e-node.sh [hunk 1]
```diff
@@ -148,7 +148,7 @@ else
 
   # Do not use any network plugin by default. User could override the flags with
   # test_args.
-  test_args='--kubelet-flags="--network-plugin= --network-plugin-dir=" '$test_args
+  test_args='--kubelet-flags="--network-plugin= --cni-bin-dir=" '$test_args
```
Comment (supereagle) [inline hack/make-rules/test-e2e-node.sh]: Can both `--network-plugin=` and `--cni-bin-dir` be set through `test_args` at the same time? A bug here?

Comment (freehan) [inline hack/make-rules/test-e2e-node.sh]: I think this is for local testing where no network plugin is used. But it should be able to do it without specifying the network-plugin flag. @Random-Liu confirm?

**Comment by supereagle** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by supereagle** [comment]
Have rebased to fix the conflict with PR #44061.

@k8s-bot ok to test

**Comment by supereagle** [comment]
@k8s-bot pull-kubernetes-federation-e2e-gce test this

**Comment by supereagle** [comment]
@k8s-bot ok to test

**Comment by supereagle** [comment]
@freehan  @Random-Liu  Could you help to approve again as this PR has been rebased.

File: hack/make-rules/test-e2e-node.sh (hunk 1 above)
Comment (supereagle) [inline hack/make-rules/test-e2e-node.sh]: I think this has no effect on the tests. If it really has error, I will fix it in another PR, no need to block this PR.

**Comment by freehan** [comment]
@Random-Liu for /approve

**Comment by supereagle** [comment]
@k8s-bot pull-kubernetes-e2e-kops-aws test this

**Comment by supereagle** [comment]
Kindly ping @Random-Liu @zmerlynn  for /approve

**Comment by supereagle** [comment]
Could this be in v1.7? Otherwise we will need to wait one more release to remove this flag. @freehan

**Comment by freehan** [comment]
@Random-Liu for /approve

**Comment by freehan** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by supereagle** [comment]
@k8s-bot pull-kubernetes-e2e-gce-etcd3 test this

**Comment by supereagle** [comment]
@freehan  Will this PR be cherry-picked into v1.7?  It has been labeled `cherrypick-candidate`.

**Comment by freehan** [comment]
I think it has been cherry-picked

**Comment by supereagle** [comment]
@freehan  yup, thanks.

**Pull request merged**